code/modules/power/chem_combustion.dm

```diff
@@ -0,0 +1,394 @@
+
+		var/turf/simulated/T = get_turf(src)
+		if (istype(T))
+			var/datum/gas_mixture/payload = new /datum/gas_mixture
+			payload.carbon_dioxide = src.atmos_drain_rate * average_volatility * src.output_multiplier
+			payload.temperature = T20C
+			T.assume_air(payload)
+
+		if (src.inlet_tank)
+			src.inlet_tank.air_contents.remove(src.atmos_drain_rate * average_volatility * src.output_multiplier)
+		else
+			T.air.remove(src.atmos_drain_rate * average_volatility * src.output_multiplier)
+
+		src.fuel_tank.reagents.remove_any(src.fuel_drain_rate * src.output_multiplier)
+
+		src.UpdateIcon()
+		src.updateDialog()
+
+	proc/get_average_volatility(datum/reagents/R)
+		if (!R || !R.total_volume)
+			return
+
+		var/average = 0
+		for (var/reagent_id in R.reagent_list)
+			var/datum/reagent/current_reagent = R.reagent_list[reagent_id]
+			if (current_reagent)
+				switch(reagent_id)
+					if ("foof", "kerosene", "nitrotri_dry", "dbreath")
+						average += (FUEL_QUALITY_VERY_HIGH * current_reagent.volume)
+
+					if ("blackpowder", "phlogiston", "napalm_goo", "sorium", "firedust")
+						average += (FUEL_QUALITY_HIGH * current_reagent.volume)
+
+					if ("fuel", "acetone")
+						average += (FUEL_QUALITY_AVERAGE * current_reagent.volume)
+
+					if ("oil", "butter", "diethylamine", "ethanol")
+						average += (FUEL_QUALITY_LOW * current_reagent.volume)
+
+					if ("plasma", "magnesium", "phosphorus")
+						average += (FUEL_QUALITY_VERY_LOW * current_reagent.volume)
+
+					else
+						var/datum/reagent/fooddrink/alcoholic/current_alcoholic_reagent = current_reagent
+						if (istype(current_alcoholic_reagent) && current_alcoholic_reagent.alch_strength >= 0.5)
+							average += 3
+
+		if (!average)
+			return 0
+
+		return average / R.total_volume
+
+#undef FUEL_QUALITY_VERY_HIGH
+#undef FUEL_QUALITY_HIGH
+#undef FUEL_QUALITY_AVERAGE
+#undef FUEL_QUALITY_LOW
+#undef FUEL_QUALITY_VERY_LOW
+
+	proc/start_engine()
+		if (!src.active)
+			if (!src.ready_to_start())
+				return 1
+
+			src.active = 1
+			src.UpdateIcon()
+			src.updateDialog()
+
+			if (!ON_COOLDOWN(src, "tractor", 2 SECOND))
+				playsound(src.loc, "sound/machines/tractorrev.ogg", 40, pitch=2)
+
+			return 0
+
+	proc/stop_engine()
+		if (src.active)
+			src.active = 0
+			src.UpdateIcon()
+			src.updateDialog()
+
+			if (!ON_COOLDOWN(src, "tractor", 2 SECOND))
+				playsound(src.loc, "sound/machines/tractorrev.ogg", 40, pitch=2)
+
+			return 0
+
+	proc/ready_to_start()
+		if (!anchored || !src.fuel_tank)
+			return 0
+
+		if (!src.get_average_volatility(src.fuel_tank.reagents) || !src.check_available_oxygen())
+			return 0
+
+		return 1
+
+	// Returns the concentration of oxygen in the available gas_mixture
+	proc/check_available_oxygen()
+		if (src.inlet_tank)
+			return src.check_tank_oxygen(src.inlet_tank)
+
+		var/turf/simulated/T = get_turf(src)
+		if (!istype(T))
+			return 0
+
+		if (!T.air || T.air.oxygen <= 0)
+			return 0
+
+		return T.air.oxygen / TOTAL_MOLES(T.air)
```

Comment:
You seem to be using this to check if there's oxygen to use, but are actually returning the ratio of the oxygen concentration. If there's the tiniest bit of oxygen in the air, this will still run. Was that the intention?

Reply:
The intention was to adjust the generators output based on the concentration of oxygen which is why this was needed, also the PACMAN use a similar way to check the tanks, which works for them in game. The generator will remove oxygen until there is 0 left causing the generator to run out.